ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -139,7 +139,7 @@ withNetworkNode
   => NetworkInterface ptcl Socket.AddrInfo IO r
   -> (NetworkNode Socket.AddrInfo IO r -> IO t)
```

Comment:
An idea for easily allowing servers that run indefinitely: make this an optional parameter? When it's `Nothing`, use `STM.retry` as the main action to `Server.run`.

Reply:
Or alternatively, why not just export all of the parts of `withNetworkNode`? Then a user can quite easily use `Server.run` directly with their own main

Reply:
Thanks, that's a nice idea.

Reply:
You can't do `atomically retry` because it throws `thread blocked indefinitely in an STM transaction`.

Reply:
This was resolved on slack but for the record:

> That doesn't mean we can't use `STM.retry` in the server though, because the main loop samples a `TVar` before calling the main action, in this case `const STM.retry`, so the RTS won't be able to judge a deadlock (the `TVar` is still live)

Reply:
Indeed, thanks @avieth .